macrosynergy/download/dataquery.py

```diff
@@ -285,540 +317,629 @@ def _get_token(self) -> str:
-                dq_resource_id=kwargs.pop("resource_id", OAUTH_DQ_RESOURCE_ID),
+            # ensure that we have a client_id and client_secret
+            try:
+                for k in ["client_id", "client_secret"]:
+                    assert k in kwargs, f"{k} must be provided."
+            except AssertionError as e:
+                raise ValueError(e)
+            except Exception as e:
+                raise e
+
+            self.access_method: OAuth = OAuth(
+                client_id=kwargs["client_id"],
+                client_secret=kwargs["client_secret"],
+                base_url=OAUTH_BASE_URL,
+                token_url=OAUTH_TOKEN_URL,
+                dq_resource_id=OAUTH_DQ_RESOURCE_ID,
                 proxy=self.proxy,
             )
+
         else:
-            self.access: CertAuth = CertAuth(
-                username=kwargs.pop("username"),
-                password=kwargs.pop("password"),
-                crt=kwargs.pop("crt"),
-                key=kwargs.pop("key"),
-                base_url=kwargs.pop("base_url", CERT_BASE_URL),
+            # ensure that we have a username and password, crt and key
+            try:
+                for k in ["username", "password", "crt", "key"]:
+                    assert k in kwargs, f"{k} must be provided."
+            except AssertionError as e:
+                raise ValueError(e)
+            except Exception as e:
+                raise e
+
+            self.access_method: CertAuth = CertAuth(
+                username=kwargs["username"],
+                password=kwargs["password"],
+                crt=kwargs["crt"],
+                key=kwargs["key"],
+                base_url=CERT_BASE_URL,
                 proxy=self.proxy,
             )
 
-        self.debug: bool = debug
-        self.last_url: Optional[str] = None
-        self.status_code: Optional[int] = None
-        self.last_response: Optional[str] = None
-        self.concurrent: bool = concurrent
-        self.batch_size: int = batch_size
-
     def __enter__(self):
         return self
 
-    def __exit__(self, exc_type, exc_value, exc_traceback):
+    def __exit__(self, exc_type, exc_value, traceback):
         if exc_type:
-            logger.error(f"Execution {exc_type} with value (exc_value): {exc_value}")
-        debug_stream_handler.stream.flush()
-        debug_stream_handler.stream.seek(0)
-        self.msg_errors = debug_stream_handler.stream.getvalue().splitlines()
-        # NOTE: Don't close the stream, as it causes can issues with parent/logging modules.
-        # NOTE: DO NOT try and close/delete self or pass it to gc.collect() here.
-
-    def check_connection(self) -> Tuple[bool, dict]:
-        """Check connection (heartbeat) to DataQuery."""
-        endpoint = "/services/heartbeat"
-        js, success, msg = self.access.get_dq_api_result(
-            url=self.access.base_url + endpoint,
+            print(f"Exception: {exc_type} {exc_value}")
+        return True
```

Comment:
Why return True on exit?

Reply:
That was a mix-up. I wanted to follow exit0, exit1 ideas to make sure the context manager returns a status code on exit.
this is useful when we don't use a context manager and do something to the effect of :
```python
dq = Interface()
exitStatus = dq.close()
if exitStatus: # in case of error follows the idea of exit(1), so would be Truem or 1.
    raise Exception
```
I've removed that now. In case of an exception, the exception will be raised. Otherwise it's up to the defaults of the context-manager and/or garbage-collector.